Enabling alias annotation checks for all operations during autograd tests

rahuln32 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46601 Enabling alias annotation checks for all operations during autograd tests**

* except excluded tests and magic methods.

https://github.com/pytorch/pytorch/issues/38731

Previously, we'd only do run these tests for inplace operations. Since this is a lot more tests, fixed these issues that came up when running them -
- Updated schema of conj() to reflect existing behaviour.
- Updated deepEquals method in check_alias_annotation.cpp to re-use the overloaded == operator. Previous implementation did not cover all types of IValues.
- Corrected the order inputs are passed in during autograd testing of 'view' & 'reshape'.
- Subbed out atn::ger with the func its aliased to, atn::outer, for testing. The alias annotation checking code doesn't handle aliased operators properly.

Differential Revision: [D24424955](https://our.internmc.facebook.com/intern/diff/D24424955/)

Commit:
Enabling alias annotation checks for all operations during autograd tests

* except excluded tests and magic methods.

https://github.com/pytorch/pytorch/issues/38731

Previously, we'd only do run these tests for inplace operations. Since this is a lot more tests, fixed these issues that came up when running them -
- Updated schema of conj() to reflect existing behaviour.
- Updated deepEquals method in check_alias_annotation.cpp to re-use the overloaded == operator. Previous implementation did not cover all types of IValues.
- Corrected the order inputs are passed in during autograd testing of 'view' & 'reshape'.
- Subbed out atn::ger with the func its aliased to, atn::outer, for testing. The alias annotation checking code doesn't handle aliased operators properly.

Differential Revision: [D24424955](https://our.internmc.facebook.com/intern/diff/D24424955/)

[ghstack-poisoned]

eellison commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -978,7 +978,7 @@ def method_tests():
         ('bmm', (M, S, M), ((M, M, S),), '', (True,)),
         ('mv', (S, M), ((M,),), '', (True,)),
-        ('ger', (S,), ((M,),)),
+        ('outer', (S,), ((M,),)),
```
This is a pretty innocuous change, but it would be nice if adding an aliased schema test didn't fail. Since the change here is not too complicated do you mind doing it as part of this PR? We can expose the https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/normalize_ops.cpp#L10 map here in an API and try to normalize the op name if we didnt fine any nodes with the original name

rahuln32 replied:
I didn't know that map existed. That's pretty sweet! Will use it.

eellison reviewed (commented):
Sweet, looks great!  I was kind of hoping we would find a bug lol. One comment on supporting aliased operators that I think would be nice to address before landing. Note that you'll need to add `aten::conj` to the backwards compatibility test here, bc the test doesnt recognize the schema as being equivalent: https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py#L37

rahuln32 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -314,7 +314,7 @@
   variants: function
 
-- func: conj(Tensor self) -> Tensor
+- func: conj(Tensor(a) self) -> Tensor(a)
```
Yes :D

rahuln32 commented on test/test_jit.py:
```diff
@@ -15723,7 +15723,7 @@ def fn(*inputs, **kwargs):
 
                 # alias annotation testing
-                if is_inplace and test_name not in EXCLUDE_SCRIPT:
+                if not is_magic_method and test_name not in EXCLUDE_SCRIPT:
```
The test was trying to call them on the Tensor. See https://www.internalfb.com/intern/testinfra/diagnostics/7599824394341892.562949968807174.1603144415/

rahuln32 commented on torch/csrc/jit/passes/utils/check_alias_annotation.cpp:
```diff
@@ -61,19 +61,11 @@ Stack deepCopy(const Stack& stack) {
-  } else if (lhs.isIntList() && rhs.isIntList()) {
-    return lhs.toIntVector() == rhs.toIntVector();
-  } else if (lhs.isTensor() && rhs.isTensor()) {
+  if (lhs.isTensor() && rhs.isTensor()) {
```
Unfortunately, it was not. '==' calls 'eq', which returns a tensor, and then attempts to convert it into a bool value by calling 'is_nonzero()'. 'is_nonzero()' was returning this exception for a bunch of the test cases - 'Boolean value of Tensor with more than one value is ambiguous' when attempting to do validation. I'm not sure why it was happening so I elected to retain the existing comparison logic in the utility class.

See :
- https://github.com/pytorch/pytorch/blob/6a6552576df182ed100f80a4da4bb1794b0c6d7c/aten/src/ATen/native/TensorCompare.cpp#L179
- https://www.internalfb.com/intern/testinfra/diagnostics/5348024594613463.844424945523256.1603155707/

rahuln32 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -586,13 +586,13 @@ def method_tests():
         ('t', (1, 2), NO_ARGS, '', (False,)),
         ('view', (S, S, S), (S * S, S), '', (False,)),
-        ('view', (S, S, S), (torch.Size([S * S, S]),), 'size', (False,)),
+        ('view', (torch.Size([S * S, S]),), (S, S, S), 'size', (False,)),
```
Not sure why either. Might be how the rest of the tests access the input variables differently?

rahuln32 commented:
Closing this in favor of https://github.com/pytorch/pytorch/pull/46635

Commit:
Update on "Enabling alias annotation checks for all operations during autograd tests"


* except excluded tests and magic methods.

https://github.com/pytorch/pytorch/issues/38731

Previously, we'd only do run these tests for inplace operations. Since this is a lot more tests, fixed these issues that came up when running them -
- Updated schema of conj() to reflect existing behaviour.
- Updated deepEquals method in check_alias_annotation.cpp to re-use the overloaded == operator. Previous implementation did not cover all types of IValues.
- Corrected the order inputs are passed in during autograd testing of 'view' & 'reshape'.
- Subbed out atn::ger with the func its aliased to, atn::outer, for testing. The alias annotation checking code doesn't handle aliased operators properly.

Differential Revision: [D24424955](https://our.internmc.facebook.com/intern/diff/D24424955/)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/normalize_ops.h:
```diff
@@ -12,5 +12,7 @@ namespace jit {
 // Currently only handles normalization of op aliases.
 TORCH_API void NormalizeOps(const std::shared_ptr<Graph>& graph);
 
+extern const std::unordered_map<Symbol, Symbol> alias_map;
```
nit: I would have some slight preference to making this an AP to get the map instead of exposing the map directly. i think the main reason is to not clutter the namespace, & that seems to be how we do it in some other files https://github.com/pytorch/pytorch/blob/cebe87fe3a66a74b6fdbb260bdfadb453444a151/torch/csrc/jit/frontend/sugared_value.cpp#L34. 

but this is a nit / personal preference, it's up to you.

rahuln32 replied:
I'd prefer to keep my changes consistent with the rest of the code base. Thanks for pointing this out. I've changed it.

eellison reviewed (approved):
Ship it!

Commit:
Update on "Enabling alias annotation checks for all operations during autograd tests"


* except excluded tests and magic methods.

https://github.com/pytorch/pytorch/issues/38731

Previously, we'd only do run these tests for inplace operations. Since this is a lot more tests, fixed these issues that came up when running them -
- Updated schema of conj() to reflect existing behaviour.
- Updated deepEquals method in check_alias_annotation.cpp to re-use the overloaded == operator. Previous implementation did not cover all types of IValues.
- Corrected the order inputs are passed in during autograd testing of 'view' & 'reshape'.
- Subbed out atn::ger with the func its aliased to, atn::outer, for testing. The alias annotation checking code doesn't handle aliased operators properly.

Differential Revision: [D24424955](https://our.internmc.facebook.com/intern/diff/D24424955/)

[ghstack-poisoned]
